// hyperclick/lib/Hyperclick.js
'use babel';
/* @flow */

var {Range} = require('atom');

type HyperclickProvider = {
  // Use this to provide a suggestion for single-word matches.
  // Optionally set `wordRegExp` to adjust word-matching.
  getSuggestionForWord?: (textEditor: TextEditor, text: string, range: Range) =>
      ?Promise<HyperclickSuggestion>;
  wordRegExp?: RegExp;

  // Use this to provide a suggestion if it can have non-contiguous ranges.
  // A primary use-case for this is Objective-C methods.
  getSuggestion?: (textEditor: TextEditor, position: Point) => ?Promise<HyperclickSuggestion>;
};

type HyperclickSuggestion = {
  // The range(s) to underline to provide as a visual cue for clicking.
  range: ?Range | ?Array<Range>;

  // The function to call when the underlined text is clicked.
  callback: () => void;
};

/**
 * Returns the text and range for the word that contains the given position.
 */
function getWordTextAndRange(
    textEditor: TextEditor,
    position: Point,
    wordRegExp?: ?RegExp): {text: string; range: Range} {
  if (!wordRegExp) {
    wordRegExp = textEditor.getLastCursor().wordRegExp();
  }

  var textAndRange = {text: '', range: new Range(position, position)};
  var buffer = textEditor.getBuffer();
  buffer.scanInRange(wordRegExp, buffer.rangeForRow(position.row), data => {
    if (data.range.containsPoint(position)) {
      textAndRange = {
        text: data.matchText,
        range: data.range,
      };
      data.stop();
    } else if (data.range.end.column > position.column) {
      // Stop the scan if the scanner has passed our position.
      data.stop();
    }
  });

  return textAndRange;
}

/**
 * Calls the given functions and returns the first non-null return value.
 */
function findTruthyReturnValue(fns): Promise {
  return new Promise((resolve, reject) => {
    var fnsLength = fns.length;
    var next = async function(index) {
      if (index === fnsLength) {
        resolve(null);
        return;
      }

      var fn = fns[index];
      var result = typeof fn === 'function' ? await fn() : null;
      if (result) {
        resolve(result);
      } else {
        next(index + 1);
      }
    };

    next(0);
  });
}

/**
 * Construct this object to enable Hyperclick in the Atom workspace.
 * Call `dispose` to disable the feature.
 */
class Hyperclick {
  constructor() {
    this._consumedProviders = [];

    this._hyperclickForTextEditors = new Set();
    this._textEditorSubscription = atom.workspace.observeTextEditors(textEditor => {
      var HyperclickForTextEditor = require('./HyperclickForTextEditor');
      var hyperclickForTextEditor = new HyperclickForTextEditor(
          textEditor,
          this._getSuggestion.bind(this));
      this._hyperclickForTextEditors.add(hyperclickForTextEditor);

      textEditor.onDidDestroy(() => {
        this._hyperclickForTextEditors.delete(hyperclickForTextEditor);
      });
    });
  }

  dispose() {
    if (this._textEditorSubscription) {
      this._textEditorSubscription.dispose();
      this._textEditorSubscription = null;
    }
    this._hyperclickForTextEditors.forEach(hyperclick => hyperclick.dispose());
    this._hyperclickForTextEditors.clear();
    this._consumedProviders = [];
  }

  consumeProvider(provider: HyperclickProvider | Array<HyperclickProvider>): void {
    if (Array.isArray(provider)) {
      this._consumedProviders = this._consumedProviders.concat(provider);
    } else {
      this._consumedProviders.push(provider);
    }
  }

  /**
   * Returns the first suggestion from the consumed providers.
   */
  _getSuggestion(textEditor: TextEditor, position: Point): Promise {
    return findTruthyReturnValue(this._consumedProviders.map(provider => {
      if (provider.getSuggestion) {
        return () => provider.getSuggestion(textEditor, position);
      } else if (provider.getSuggestionForWord) {
        return () => {
          var {text, range} = getWordTextAndRange(textEditor, position, provider.wordRegExp);
          return provider.getSuggestionForWord(textEditor, text, range);
        };
      }

      throw new Error('Hyperclick must have either `getSuggestion` or `getSuggestionForWord`')
    }));
  }
}

module.exports = Hyperclick;
